overlord/devicestate/handlers_systems.go

```diff
@@ -121,6 +132,148 @@ func purgeNewSystemSnapFiles(logfile string) error {
 }
 
+func snapsUniqueToRecoverySystem(target string, systems []*System) ([]string, error) {
+	// asserted snaps are shared by systems, figure out which ones are unique to
+	// the system we want to remove
+	requiredByOtherSystems := make(map[string]bool)
+	for _, sys := range systems {
+		if sys.Label == target {
+			continue
+		}
+
+		sd, err := seed.Open(dirs.SnapSeedDir, sys.Label)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := sd.LoadAssertions(nil, func(*asserts.Batch) error {
+			return nil
+		}); err != nil {
+			return nil, err
+		}
+
+		if err := sd.LoadMeta(seed.AllModes, nil, timings.New(nil)); err != nil {
+			return nil, err
+		}
+
+		err = sd.Iter(func(sn *seed.Snap) error {
+			if sn.ID() != "" {
+				requiredByOtherSystems[sn.Path] = true
+			}
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	targetSeed, err := seed.Open(dirs.SnapSeedDir, target)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := targetSeed.LoadAssertions(nil, func(*asserts.Batch) error {
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	if err := targetSeed.LoadMeta(seed.AllModes, nil, timings.New(nil)); err != nil {
+		return nil, err
+	}
+
+	var uniqueToTarget []string
+	err = targetSeed.Iter(func(sn *seed.Snap) error {
+		if sn.ID() != "" && !requiredByOtherSystems[sn.Path] {
+			uniqueToTarget = append(uniqueToTarget, sn.Path)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return uniqueToTarget, nil
+}
+
+func (m *DeviceManager) doRemoveRecoverySystem(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	systems, err := m.systems()
+	if err != nil {
+		return fmt.Errorf("cannot get recovery systems: %w", err)
+	}
+
+	setup, err := taskRemoveRecoverySystemSetup(t)
+	if err != nil {
+		return err
+	}
+
+	found := false
+	for _, sys := range systems {
+		if sys.Label == setup.Label {
+			found = true
+			if sys.Current {
+				return fmt.Errorf("cannot remove current recovery system: %q", setup.Label)
+			}
+			break
+		}
+	}
+
+	// if we couldn't find the system in the returned list of systems, then we
+	// might have already attempted to remove it. in that case, we should do our
+	// best effort at making sure that it is fully removed.
+
+	if found && len(systems) == 1 {
+		return fmt.Errorf("cannot remove last recovery system: %q", setup.Label)
+	}
+
+	deviceCtx, err := DeviceCtx(st, t, nil)
+	if err != nil {
+		return fmt.Errorf("cannot get device context: %w", err)
+	}
+
+	// if this task has partially run before, then we might have already removed
+	// some files required to load the being-removed system seed. to avoid this,
+	// we first check if we already have stored a list of snaps to remove
+	// (meaning, this task is being re-run). if the list isn't present, then we
+	// calculate it and store it in the task state for potential future re-runs.
+	var snapsToRemove []string
+	if err := t.Get("snaps-to-remove", &snapsToRemove); err != nil {
```

Comment:
I think I'd use a struct here instead of a bare list, to relieve the ambiguity of empty list/nil. Hypothetical case, but ` snapsUniqueToRecoverySystem` may return `nil` when there are no unique snaps, then `t.Set(.., nil)` drops the key, so nothing gets stored in the state. Then if the task is re-run you'd enter this branch again as t.Get() would return ErrNoState, so the case of no unique snaps and one where no call to snapsUniqueToRecoverySystem was made would be indistinguishable. Alternatively, snapsUniqueToRecoverySystem would need to return an empty list rather than nil.

Reply:
Good call, forgot that nil is equivalent to delete in the state.